### ISSUE TITLE
The balance parameter for MNIST

### ISSUE THREAD

**Issue by ngmq** [ganggit/mmdpnc#1]
Hi, I am attempting to reproduce the results for the MNIST dataset in your paper. After setting every other parameters similar to the paper (learning rate of SVM = 0.01, \lambda = 3), I still have not been able to achieve similar F and V scores reported in the paper, often get something much lower. I think it might be because the balance parameter C in the file dpmm_mmc.m, line 19. This parameter was not mentioned in the paper. Could you let me know what it was for MNIST to have F-score = 0.368 and V-score = 0.389? Thank you.

**Comment by ganggit** [comment]
Hi Quan,

I am glad to hear from you for your interest in our work in nonparametric
clustering. And I am a little sorry for late reply. Basically, \lambda in
the Eq. 9 is the most important parameter, which controls the radius like
mean shift. So you need to tune \lambda (for example, you can use grid
search). Also the paper also use PCA to do dimension reduction first, which
will generate a space influence how to tune \lambda. I do not know how you
do PCA and how to tune \lambda. Please let me know if I can help you
further to reproduce the results.

Thanks,
Gang

**Comment by ngmq** [comment]
Hi Gang, thank you for your reply. I am using the following Python code to read mnist and do PCA. The basic idea is to read the dataset into Numpy arrays, then use PCA from scikit-learn to perform PCA and then randomly sample 2000 data points from the original 6000 images.

```python
import os
import struct
import numpy as np
from sklearn.decomposition import PCA
from sklearn.externals import joblib
import random
random.seed(0)

def read(dataset = "training", path = "."):
    """
    Python function for importing the MNIST data set.  It returns an iterator
    of 2-tuples with the first element being the label and the second element
    being a numpy.uint8 2D array of pixel data for the given image.
    """
    fname_img = os.path.join(path, 'train-images.idx3-ubyte')
    fname_lbl = os.path.join(path, 'train-labels.idx1-ubyte')

    # Load everything in some numpy arrays
    with open(fname_lbl, 'rb') as flbl:
        magic, num = struct.unpack(">II", flbl.read(8))
        lbl = np.fromfile(flbl, dtype=np.int8)

    with open(fname_img, 'rb') as fimg:
        magic, num, rows, cols = struct.unpack(">IIII", fimg.read(16))
        img = np.fromfile(fimg, dtype=np.uint8).reshape(len(lbl), rows, cols)
        print img.shape

    ### Dimensionality Reduction using PCA
    img = img.reshape(num, rows * cols)
    pca = PCA(n_components=100)
    pca.fit(img)
    joblib.dump(pca, "pca.pkl")

    pca = joblib.load("pca.pkl")

    print(sum(pca.explained_variance_ratio_))

    img = img.reshape(num, rows * cols)
    img100 = pca.transform(img)

    ### Randomly sample 2000 data points from 6000 images
    indices = range(num)
    selected_indices = random.sample(indices, 2000)
    data_lbl = lbl[selected_indices].reshape(-1, 1)
    data_img = img100[selected_indices]

    print data_lbl.shape
    print data_img.shape

    data = np.hstack((data_lbl, data_img))
```

The code is a direct modification from https://gist.github.com/akesling/5358964. Do you think the code should produce the same dataset for clustering as in the paper?

**Comment by ganggit** [comment]
Hi Quan, 

The script should work. I make a little change to test it on MNIST test data. 
Let me know how do you set \lambda to do the clustering. Also it is better to do data alignment before clustering. 

Thanks,
Gang


import os
import struct
import numpy as np
from sklearn.decomposition import PCA
from sklearn.externals import joblib
import random
import gzip
from matplotlib import pyplot
random.seed(0)


def read(dataset="training", path="."):
    """
    Python function for importing the MNIST data set.  It returns an iterator
    of 2-tuples with the first element being the label and the second element
    being a numpy.uint8 2D array of pixel data for the given image.
    """
    fname_img = os.path.join(path, 't10k-images-idx3-ubyte.gz')
    fname_lbl = os.path.join(path, 't10k-labels-idx1-ubyte.gz')

    # Load everything in some numpy arrays
    with gzip.open(fname_lbl, 'r') as flbl:
        magic, num = struct.unpack(">II", flbl.read(8))
        buf = flbl.read(1 * num)
        labels = np.frombuffer(buf, dtype=np.uint8).astype(np.int64)

    with gzip.open(fname_img, 'r') as fimg:
        magic, num, rows, cols = struct.unpack(">IIII", fimg.read(16))
        buf = fimg.read(num*rows*cols)
        img = np.frombuffer(buf, dtype=np.uint8).astype(np.int64).reshape(
            len(labels), rows, cols)
        print(img.shape)

    # Dimensionality Reduction using PCA
    img = img.reshape(num, rows * cols)
    pca = PCA(n_components=100)
    pca.fit(img)
    joblib.dump(pca, "pca.pkl")

    pca = joblib.load("pca.pkl")

    print(sum(pca.explained_variance_ratio_))

    img = img.reshape(num, rows * cols)
    # show a example here
    a = img[1].reshape(rows, cols)
    pyplot.imshow(a)
    pyplot.show()
    img100 = pca.transform(img)

    # Randomly sample 2000 data points from 6000 images
    indices = range(num)
    selected_indices = random.sample(indices, 2000)
    data_lbl = labels[selected_indices].reshape(-1, 1)
    data_img = img100[selected_indices]

    print(data_lbl.shape)
    print(data_img.shape)

    data = np.hstack((data_lbl, data_img))
    return data


data = read("MNIST")

**Comment by ngmq** [comment]
Thank you very much, I am glad to know that the data pre-processing is correct. I search for \lambda by trying out different values from 0.01 to 10 with varying step sizes change for different interval of values, for example the step size is 0.04 for values from 0.01 to 1.0 and 0.5 for values from 1.0 to 10.0. None of the tested \lambda produces expected results. 

Today I attempted to reproduce the results of a much smaller dataset, the Wdbc dataset, assuming that it would be easier to reproduce the results there since there is no randomness in the data. The reported results are F = 0.85 and V = 0.57. So far I also have not been able to get anything near those results. This dataset has 569 data points of 2 groundtruth classes. The resulting clusters generated from the MATLAB code in this repository often have one big cluster of size 500 and three to five smaller clusters of size 26, 4, and 1. The F and V scores are approximately 0.694 and 0.131.

Could you please clarify some of the following points?

1) In the MATLAB code, the \lambda parameter is the value 0.5 in line 89, file dpmm_mmc.m. For wdbc (not MNIST), this value will be 3 as reported in the paper. Is that correct?

2) The comments in line 19 says that C = 3 for wdbc. Is this comment still valid or should I set C = 1 since the paper did not mention this parameter? For wdbc, when I set C = 1 the clustering is degenerate i.e. it contains only 1 cluster of all data points.

3) The F-score is computed according to table I and II of the paper by Achtert, 2012: Evaluation of Clusterings – Metrics and Visual Support, is that correct? That is, F = 2a / (2a + b + c) where a is the number of pairs in both clusterings, b and c are the number of pairs in one clustering but not in the other.

4) The V-score is computed according to the paper by Andrew Rosenberg and Julia Hirschberg, 2007. V-Measure: A conditional entropy-based external cluster evaluation measure. Is that correct? Did you use the function sklearn.metrics.v_measure_score from scikit-learn?

I understand that it's been 2 years since the paper is published, so it might be difficult for you to retrieve all these details. Any help is appreciated!

P/S: the procedure for running on wdbc that I used:

1. Download the wdbc data from https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data.
2. Remove the first two columns of every row (they are ID and label).
3. Call the clustering procedure:
```
data = csvread('wdbc.data');
params = dpmm_mmc(data, 100);
```
4. Extract clusters from `params(end).classes`.
5. Compare the extracted clusters with groundtruth clusters using F-score and V-score computed above.

**Comment by ganggit** [comment]
Hi Quan,

I uploaded mnist_test2.m and digits.mat (MNIST) to the repo. You can run mnist_test2 and see what you can get. In the dpmm_mmc.m, you can set C = 3->10 (grid search or other), I used C=4.25 and run 1 iteration, not test much. Let me know whether you can reproduce the result or not. 

Thanks,
Gang

**Comment by ganggit** [comment]
Hi Quan,

Do you run the script that I uploaded to the repo? Does it work? If you can
reproduce the result, then I will close this issue. In addition, if you are
interested in research, we can discuss more.

Thanks,
Gang

**Comment by ngmq** [comment]
Hi Gang,

Sorry for the late reply. Thank you for the scripts. I have been working on a paper for an upcoming deadline. I will run the script and report the results next week.

**Comment by ngmq** [comment]
Hi Gang,

I was able to run your scripts today. Unfortunately so far I still have not been able to reproduce the results. When I set lambda = 3 and C = 1.8 or C = 4.25, the resultant cluster has only 1 big cluster containing all 2000 data points, in this case F-score = 0.18 and V-score = 0. If I reduce lambda to smaller like 0.2 or so, I get more sensible clustering result with 4 or 5 clusters, the F and V scores change accordingly. Overall, as you said, lambda and C have to be tuned by some hyper-parameter searching strategy, and I figure that it might take a while to find out which values work. I think it would not be very productive for me to reinvent the wheel here, so I will stop working on reproducing the results in the paper.

Thank you very much for your support on this issue. I learn a lot from your code.
